fix(react-bindings): multiple client async start handling while loading

nadavshatz opened the pull request.

nadavshatz commented:
One other thought I had was maybe we should move the loading being set into a useEffect. It might be consistent with react principles and safer for future changes. 

Let me know if you agree and I can change it accordingly. 

Didn't want to make a larger change without discussing first.

daniel-statsig commented:
Hey @nadavshatz, thanks for the PR. I don't quite follow what it is solving however.

`initializeAsync` is making a network request and then setting `isLoading` to false once it is resolved.

My understanding of the changes in this PR is that if a second call to `useStatsigInternalClientFactoryAsync` is made, it will just immediately resolve and set isLoading to false, regardless if the network request is still in flight.

nadavshatz commented:
For some reason on a second call or rerender (not sure the reason I just see a debugger being called again inside the use memo) then it just returns since the ref exists but the loading isn't being set since the code isn't reached for that instance. Maybe each call creates another instance of the is loading state.

nadavshatz commented:
After some more research it is indeed considered bad practice to modify `state` while in a useMemo.

Working with ChatGPT even provides this useEffect version which is much much safer.

I updated my PR to it and added <StrictMode> to the test which actually fails with the old implementation.

nadavshatz commented:
lastly - we can use this if we don't want the strict mode wrapper.
https://testing-library.com/docs/react-testing-library/api/#reactstrictmode

I just didn't want to change a global setting.

daniel-statsig commented on packages/react-bindings/src/useStatsigInternalClientFactoryAsync.ts:
```diff
@@ -14,24 +14,25 @@ export function useStatsigInternalClientFactoryAsync<T extends StatsigClient>(
+
+  useEffect(() => {
+    // already initializing or initialized
+    if (clientRef.current.loadingStatus !== 'Uninitialized') {
```
Can you better outline what it is you are trying to solve? This change means that in strict mode, we will immediately render, regardless of the SDKs status.

Adding this logic means that we will just set isLoading to false on the re-render as its loadingStatus will be "Loading" and therefore pass this check.

daniel-statsig replied:
This logic is intended to prevent multiple SDK clients/network requests being created.

In React Strict mode, a re-render occurs, blowing away the previous state.
The logic here is supposed to create the client on the first render, then reuse it on the second.

nadavshatz replied:
Correct. 
As far as I could tell, the old logic tried to do the same.

daniel-statsig replied:
This is incorrect. The loadingStatus goes from "Uninitialized" -> "Loading" -> "Ready". Do this here means that we would potentially `setIsLoading(false)` when loading status is "Loading"

daniel-statsig commented:
Closing as this would break loading states
